File: python/python/ch4/impure_and_pure.py
from ..ch1.db import StudentDB
import toolz as tz
import pydash

def impure_and_pure():
    find = lambda db, id: list(tz.filter(lambda student: student['ssn'] == id, db.students))[0]
    
    def find_object_curry(db, id):
        obj = find(db, id)
        if obj == None:
            raise NameError(f'ID가 [{id}]인 객체는 없습니다.')
          
        return obj
    find_object = pydash.curry(find_object_curry)
    
    db = StudentDB()
    find_student = find_object(db)
    full_info = lambda student: f'{student["ssn"]}, {student["firstname"]}, {student["lastname"]}'
    append = pydash.curry(lambda error, info: print('error' if error == True else 'info', info))
    show_student = tz.compose(
      append(False),
      full_info,
      find_student
    )
    
    show_student('444-44-4444')

File: javascript/ch4/impureAndPure.js
const { StudentDB } = require('../ch1/db');
const fp = require('lodash/fp');

function impureAndPure() {
  const find = (db, id) => db.students.filter((student) => student.ssn === id)[0];
  const findObject = fp.curry((db, id) => {
    const obj = find(db, id);
    if (obj == null) {
      throw new Error(`ID가 [${id}]인 객체는 없습니다.`);
    }

    return obj;
  });

  const db = new StudentDB();
  const findStudent = findObject(db);
  const fullInfo = ({ firstname, lastname, ssn }) => 
                      `${ssn}, ${firstname}, ${lastname}`;
  const append = fp.curry((error, info) => console.log(error ? 'error' : 'info', info));

  const showStudent = fp.compose(
    append(false),
    fullInfo,
    findStudent
  );

  showStudent('444-44-4444');
}

module.exports = {
  impureAndPure
}